modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/ExternalResultHashIndex.java

```diff
@@ -0,0 +1,559 @@
+        return entry.isEmpty() || entry.isRemoved() ? null : entry;
+    }
+
+    /**
+     * Reads hash map entry from the underlying file.
+     *
+     * @param slot number of the entry to read. {@code Null} means sequential reading.
+     * @return Byte buffer with data.
+     */
+    private Entry readEntryFromIndexFile(long slot) {
+        return readEntryFromIndexFile(slot, fileIo);
+    }
+
+    /**
+     * Reads hash map entry from the underlying file.
+     *
+     * @param slot number of the entry to read. {@code -1} means sequential reading.
+     * @param fileCh File to read from.
+     * @return Byte buffer with data.
+     */
+    private Entry readEntryFromIndexFile(long slot, FileIO fileCh) {
+        try {
+            if (slot != -1)
+                gotoSlot(slot);
+            else
+                slot = fileCh.position() / Entry.ENTRY_BYTES;
+
+            reusableBuff.clear();
+
+            synchronized (this) {
+                checkCancelled();
+
+                fileCh.readFully(reusableBuff);
+            }
+
+            reusableBuff.flip();
+
+            int hashCode = reusableBuff.getInt();
+
+            // We store addr value incremented by 1 to distinguish 0 address from the empty address field.
+            // On the disk 0 means empty, 1 means zero, y means y-1.
+            long addr = reusableBuff.getLong() - 1;
+
+            return new Entry(hashCode, addr, slot);
+        }
+        catch (IOException e) {
+            U.closeQuiet(this);
+
+            throw new IgniteException("Failed to read query result the from spill idx file. [slot=" + slot + ']', e);
+        }
+    }
+
+    /**
+     * Writes hash map entry to file.
+     *
+     * @param slot Entry slot.
+     * @param hashCode Hash code.
+     * @param addr Row address.
+     */
+    private void writeEntryToIndexFile(long slot, int hashCode, long addr) {
+        try {
+            gotoSlot(slot);
+
+            reusableBuff.clear();
+            reusableBuff.putInt(hashCode);
+
+            // We store addr value incremented by 1 to distinguish 0 address from the empty address field.
+            // On the disk 0 means empty, 1 means zero, y means y-1.
+            reusableBuff.putLong(addr + 1);
+            reusableBuff.flip();
+
+            synchronized (this) {
+                checkCancelled();
+
+                fileIo.writeFully(reusableBuff);
+            }
+        }
+        catch (IOException e) {
+            U.closeQuiet(this);
+
+            throw new IgniteException("Failed to write intermediate query result to the spill file.", e);
+        }
+    }
+
+    /**
+     * Sets arbitrary file position.
+     *
+     * @param slot Position to set.
+     */
+    private void gotoSlot(long slot) {
+        try {
+            synchronized (this) {
+                checkCancelled();
+
+                fileIo.position(slot * Entry.ENTRY_BYTES);
+            }
+        }
+        catch (Exception e) {
+            U.closeQuiet(fileIo);
```

Comment:
close() method removes file, but in case of exception fileIO is closed without file deletion.
Why we close fileIO, but keep file on disk?

Reply:
Replaced with `U.closeQuiet(this);`